folium/folium.py

```diff
@@ -550,32 +553,39 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
                             if data_min > 0 else -1)
                 data_max = (data_max if data_max > 0 else 0
                             if data_max < 0 else 1)
-            data_min, data_max = (1.01*data_min-0.01*data_max,
-                                  1.01*data_max-0.01*data_min)
-            nb_class = 6
-            color_domain = [data_min+i*(data_max-data_min)*1./nb_class
-                            for i in range(1+nb_class)]
+            nb_bins = 6
+            color_domain = list(np.linspace(data_min, data_max, nb_bins + 1))
         else:
             color_domain = None
 
         if color_domain and key_on is not None:
+            nb_bins = len(color_domain) - 1
             key_on = key_on[8:] if key_on.startswith('feature.') else key_on
-            color_range = color_brewer(fill_color, n=len(color_domain))
+            color_range = color_brewer(fill_color, n=nb_bins)
 
             def get_by_key(obj, key):
                 return (obj.get(key, None) if len(key.split('.')) <= 1 else
                         get_by_key(obj.get(key.split('.')[0], None),
                                    '.'.join(key.split('.')[1:])))
 
             def color_scale_fun(x):
```

Comment:
Here, I decided to mimic the previous behavior described in the comment below in order not to introduce any braking change:

```python
# we consider that values outside the color 
# should be affected to the first (or last) color bin.
```

However, doing so is actually producing misleading maps (for example, if we force the color scale do go from 1 to 10 with step 1 and we have -60 as a value, it would map to the same color as the bin 1 to 2 without any warning).
It might be a good idea to raise a warning in such circumstances or just to forbid this behavior (that's what I would do).
I didn't choose the last option as it was causing unit tests to fails:

Some notebooks use `branca` `split_six` functions which can produce *weird* intervals, for example:
```python
import pandas as pd
from folium import branca
threshold_scale = branca.utilities.split_six(series=pd.Series([1,2,2,3,3,3,3,3,3,4]))
# produces (as expected): [1.0, 3.0, 3.0, 3.0, 3.0]
```
Therefore, the `4` we would like to map to a color would be outside of the `threshold_scale`.

Reply:
Good that you point this out. We have to make a decision here. I'm listing the options:

1. Let values outside of the range use the min or max value (current behavior).
  * This makes sense if you have a few outliers you don't want to stretch out your range for.
  * No time to filter out outliers, get stuff done quickly.
2. Raise a UserWarning when values fall outside of the range.
  * Sort of compromise between the current behavior and raising an error.
  * Numpy does this a lot, which results in people figuring out how to disable/ignore them.
  * Kind of annoying (subjective, I know :p )
3. Raise an exception when values fall outside of the range.
  * Having a value lie outside of the range is a mistake and we don't do mistakes.
  * Force the user to give either a good range or remove outliers.
4. Give the values outside the range the same color as `nan_fill_color`.
  * This does show they are not in the range, but it makes them indistinguishable from nan values.
5. Give the values outside the range a special color and add a new argument for it.
  * Makes it easy to see which values fall outside the range.
  * Adds complexity, another argument to check out and set.

I'm currently leaning towards raising an exception. Many users may not pass a custom range, so they will not encounter this issue. If a user passes a custom range, that means he or she is advanced enough to also be able to properly preprocess the data and deal with outliers.

On the other hand I like how folium is hackable. If someone wants to deal with their outliers by cheating a bit, who are we to judge. But then maybe a UserWarning would be appropriate.

So after writing this comment I think I would lean towards leaving it as is and add a UserWarning. Do you want to make the case for another option? I'd like to hear your opinion. We can change the tests based on what we decide.

Reply:
Let's raise a `ValueError` then, with a useful message. We'll learn our users some proper preprocessing ;) Better to be explicit than have it fail silently.